README.md
<img alt="Gi-SoftWare" src="https://raw.githubusercontent.com/raix/push/master/docs/logo.png" width="20%" height="20%">

raix:push Push notifications
============================
[![Build Status](https://travis-ci.org/raix/push.svg?branch=master)](https://travis-ci.org/raix/push)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)

> Push notifications for cordova (ios, android) browser (Chrome, Safari, Firefox) - One unified api on client and server.

Status:
* [x] APN iOS
* [x] GCM/FCM Android
* [x] APN Safari web push (partially implemented)
* [x] GCM Chrome OS (partially implemented)
* [x] Firefox OS (partially implemented)
* [ ] BPS Blackberry 10
* [ ] MPNS Windows phone 8
* [ ] MPNS Windows 8
* [ ] ADM Amazon Fire OS
* [ ] Meteor in app notifications

## Install
```bash
  $ meteor add raix:push
  $ meteor add cordova:cordova-plugin-device@1.1.5
  $ meteor add cordova:phonegap-plugin-push@1.5.2
  # Note: you probably want to adjust the version numbers to the latest versions of the packages
```

## Getting started
Depending on the platforms you want to work with you will need some credentials or certificates.
* [Android](docs/ANDROID.md)
* [iOS](docs/IOS.md)

Have a look at the [Basic example](docs/BASIC.md)

Theres a good walkthrough by [Arthur Carabott](https://medium.com/@acarabott/meteor-native-ios-push-notifications-heroku-raix-push-cordova-213f486c4e6d#.akrtpzmi7)

Read the [raix:push Newbie Manual](https://github.com/raix/push/wiki/raix:push-Newbie-Manual) by [@harryward](https://github.com/harryward)

Or check out the [DEMO](https://github.com/elvismercado/meteor-raix-push-demo) by [@elvismercado](https://github.com/elvismercado)
(This example uses the deprecated config.push.json)

Example code for [sound](https://github.com/raix/push/issues/9#issuecomment-216068188) *(todo: add in docs)*

Note:
Version 3 uses the cordova npm plugin [phonegap-plugin-push](https://github.com/phonegap/phonegap-plugin-push#pushnotificationinitoptions)

Note:
Some of the documentation is outdated, please file an issue or create a pull request - same if you find a bug or want to add tests
## Development

Watch the project [progress](https://github.com/raix/push/projects/1) for status or join development

## Config

Use the `Push.Configure` function on client and server.

### Client

For example in `Meteor.startup()` block of main.js

```js
Push.Configure({
  android: {
    senderID: 12341234,
    alert: true,
    badge: true,
    sound: true,
    vibrate: true,
    clearNotifications: true
    // icon: '',
    // iconColor: ''
  },
  ios: {
    alert: true,
    badge: true,
    sound: true
  }
});
```

Additionally you have to touch `mobile-config.js`
```js
App.configurePlugin('phonegap-plugin-push', {
  SENDER_ID: 12341234
});
```
*This is due to changes in the cordova plugin itself*

### Server

For example in `Meteor.startup()` block of main.js

```js
Push.Configure({
  apn: {
    certData: Assets.getText('apnDevCert.pem'),
    keyData: Assets.getText('apnDevKey.pem'),
    passphrase: 'xxxxxxxxx',
    production: true,
    //gateway: 'gateway.push.apple.com',
  },
  gcm: {
    apiKey: 'xxxxxxx',  // GCM/FCM server key
  }
  // production: true,
  // 'sound' true,
  // 'badge' true,
  // 'alert' true,
  // 'vibrate' true,
  // 'sendInterval': 15000, Configurable interval between sending
  // 'sendBatchSize': 1, Configurable number of notifications to send per batch
  // 'keepNotifications': false,
//
});
```
*Note: `config.push.json` is deprecating*

## Common API
```js
    // Push.debug = true; // Add verbosity

    Push.send({
        from: 'push',
        title: 'Hello',
        text: 'world',
        badge: 1, //optional, use it to set badge count of the receiver when the app is in background.
        query: {
            // Ex. send to a specific user if using accounts:
            userId: 'xxxxxxxxx'
        } // Query the appCollection
        // token: appId or token eg. "{ apn: token }"
        // tokens: array of appId's or tokens
        // payload: user data
        // delayUntil: Date
    });
```
*When in secure mode the client send features require adding allow/deny rules in order to allow the user to send push messages to other users directly from the client - Read more below*

## Client API
```js
    Push.id(); // Unified application id - not a token
    Push.setBadge(count); // ios specific - ignored everywhere else
    Push.enabled(); // Return true or false
    Push.enabled(false); // Will disable notifications
    Push.enabled(true); // Will enable notifications (requires a token...)
```

## Security allow/deny send
This package allows you to send notifications from the server and client. To restrict the client or allowing the client to send use `allow` or `deny` rules.

When a client calls send on Push, the Push's allow and deny callbacks are called on the server to determine if the send should be allowed. If at least one allow callback allows the send, and no deny callbacks deny the send, then the send is allowed to proceed.

```js
    Push.allow({
        send: function(userId, notification) {
            return true; // Allow all users to send
        }
    });

    // Or...
    Push.deny({
        send: function(userId, notification) {
            return false; // Allow all users to send
        }
    });
```

## Meteor Methods

### raix:push-update

Stores a token associated with an application and optionally, a userId.

**Parameters**:

*options* - An object containing the necessary data to store a token. Fields:
* `id` - String (optional) - a record id for the Application/Token document to update. If this does not exist, will return 404.
* `token` - Object - `{ apn: 'TOKEN' }` or `{ gcm: 'TOKEN' }`
* `appName` - String - the name of the application to associate the token with
* `userId` - String (optional) - the user id so associate with the token and application. If none is included no user will be associated. Use `raix:push-setuser` to later associate a userId with a token.

**Returns**:

*recordId* - The id of the stored document associating appName, token, and optionally user in an object of the form:

```
{
  result: 'recordId'
}
```

### raix:push-setuser

Associates the current users ID with an Application/Token record based on the given id.

**Parameters**:

*id* - String - The ID of the Application/Token record

### raix:push-metadata

Adds metadata to a particular Application/Token record.

**Parameters**

*data* - Object containing the following fields:
* `id` - String - the ID of the Application/Token record to update
* `metadata` - Object - The metadata object to add to the Application/Token document

## More Info


For more internal or advanced features read [ADVANCED.md](docs/ADVANCED.md)

Kind regards

Morten (aka RaiX)


CHANGELOG.md
# Changelog

## vCurrent

- add-support-to-title-in-ios

- allow-apn-category-field

## [v2.6.0] (https://github.com/raix/push/tree/v2.6.0)
#### 28/01/15 by Morten Henriksen
- Respect resume - if paused wait until resume before triggering start up events

- make startup event stable

- enabled "alert" should not force alert - use message event instead

- make sure set badge runs when device is ready

- Use updated pushplugin

- add raix:cordova as a weak dependency

- Allow the client create doc with id instead of throwing error

## [master-workpop-pr] (https://github.com/raix/push/tree/master-workpop-pr)
#### 27/01/15 by Morten Henriksen
## [master-foo] (https://github.com/raix/push/tree/master-foo)
#### 27/01/15 by Morten Henriksen
- check that token is not undefined

- *Merged pull-request:* "raix:push-update -> server generates record id rather than client" [#32](https://github.com/raix/push/issues/32) ([alexcorre](https://github.com/alexcorre))

- fix readme italics markdown

- update to 2.6.0 for breaking change + change name back to raix:push

- bump to 2.5.4

- Make push-update return 404 if ID is included that does not yet exist

- remove another extra comma

- Add meteor method docs to README.md

- Merge branch 'master' of github.com:raix/push

- *Merged pull-request:* "Docupdates" [#29](https://github.com/raix/push/issues/29) ([funkyeah](https://github.com/funkyeah))

- remove extra comma

- bump version to 2.5.3

- fix issue with push-update

- .idea to gitignore and update version

- raix:push-update no longer requires id, generates if none included

- raix:push-update remove required but unused metadata field from options

- grammar and formatting changes to Push.send

- more details on the flexibility of Push.send

- Push.send advanced documentation updates

- Gi-Software push logo size

- add logo

- Add Gi-Software logo

Patches by GitHub users [@alexcorre](https://github.com/alexcorre), [@funkyeah](https://github.com/funkyeah).

## [v2.5.1] (https://github.com/raix/push/tree/v2.5.1)
#### 04/01/15 by Morten Henriksen
- Add updated iframe code for consistency

- Add iOS startup workaround for stable events

- Change payload behaviour - Modify payload when sending instead of storing

## [v2.5.0] (https://github.com/raix/push/tree/v2.5.0)
#### 03/01/15 by Morten Henriksen
- Improve the payload object, this is now the same on both iOS and Android and will preserve types via ejson

- add doc about open

- wrap in device ready - dont trust Meteor.startup

- Add a new `open` flag marking if the note triggered the app to open

- add more test alerts

- exclude aps from the payload

## [v2.4.1] (https://github.com/raix/push/tree/v2.4.1)
#### 03/01/15 by Morten Henriksen
- Fixed getting userId on the server throws error

## [v2.4.0] (https://github.com/raix/push/tree/v2.4.0)
#### 03/01/15 by Morten Henriksen
- Refactor Push.init to Push.Configure and add call once check

## [v2.3.3] (https://github.com/raix/push/tree/v2.3.3)
#### 03/01/15 by Morten Henriksen
- just sound location

- Add better Push.debug verbosity

## [v2.3.2] (https://github.com/raix/push/tree/v2.3.2)
#### 29/12/14 by Morten Henriksen
- Dont try to play empty sound file names

- only prefix sound if text found

## [v2.3.1] (https://github.com/raix/push/tree/v2.3.1)
#### 29/12/14 by Morten Henriksen
- 2.3.1 - added payload option

- fix default calback settings

- add payload in event

- add commented alerts for debugging

- clean up foreground code and keep coldstart

- document payload

- check and add payload

## [v2.3.0] (https://github.com/raix/push/tree/v2.3.0)
#### 29/12/14 by Morten Henriksen
- add type in events

- 2.3.0 Added vibrate and better support for sound/alerts and a unified api

- refactoring of the server send api

- add limited documentation for default event listeners

- add vibrate

- have badge and sound added only if actually set

- add badge and sound

- correct default alert and add vibrate

- only emit alert events when in foreground

- Unify the notification message object and event model

- Add basis code for apn to know about app status

- Dont emit badge event when user sets badge


## [v2.2.1] (https://github.com/raix/push/tree/v2.2.1)
#### 29/12/14 by Morten Henriksen
- Add the passphrase... ups

- use meteor startup

- Remove unused reference

- *Fixed bug:* "Sending Push Message From Browser Console Requires Logged In User" [#11](https://github.com/raix/push/issues/11)

- added more debug details

- update change log

## [v2.2.0] (https://github.com/raix/push/tree/v2.2.0)
#### 28/12/14 by Morten Henriksen
- Add support for config file

- Refactor files into folders

- Update change log

## [v2.1.3] (https://github.com/raix/push/tree/v2.1.3)
#### 27/12/14 by Morten Henriksen
- Use the matchToken in common

- Correct createdBy match - allow null

- Added a basic example

## [v2.1.2] (https://github.com/raix/push/tree/v2.1.2)
#### 27/12/14 by Morten Henriksen
- Thanks to @adam-hanna

## [v2.1.1] (https://github.com/raix/push/tree/v2.1.1)
#### 26/12/14 by Morten Henriksen
- Add more debug verbosity

- Fix android sends - convert string to array of strings

## [v2.1.0] (https://github.com/raix/push/tree/v2.1.0)
#### 26/12/14 by Morten Henriksen
- Correct documentation

- Add allow and deny rules to documentation

- Ignore setBadge on the server for now

- Add the throttled send

- Add secure client send push notification

- Refactor appId to appName

- add more credits

- fix docs

- Add more documentation - all for free

- Correct details about client send

- move setBadge to client code for now

## [v2.0.11] (https://github.com/raix/push/tree/v2.0.11)
#### 26/12/14 by Morten Henriksen
- Clean up docs and hide more advanced features

- pushId in options is now deprecated - use gcm.projectNumber, gcm.apiKey, websitePushId instead - Breaking change!

- have a general production option for Push.init

- We init apn feedback pr. default

- add more debug verbosity


push.server.js
/*
  A general purpose user CordovaPush
  ios, android, mail, twitter?, facebook?, sms?, snailMail? :)

  Phonegap generic :
  https://github.com/phonegap-build/PushPlugin
 */

// getText / getBinary

Push.setBadge = function(id, count) {
    // throw new Error('Push.setBadge not implemented on the server');
};

Push.init = function(options) {
    var self = this;
    // https://npmjs.org/package/apn

    // After requesting the certificate from Apple, export your private key as a .p12 file and download the .cer file from the iOS Provisioning Portal.

    // gateway.push.apple.com, port 2195
    // gateway.sandbox.push.apple.com, port 2195

    // Now, in the directory containing cert.cer and key.p12 execute the following commands to generate your .pem files:
    // $ openssl x509 -in cert.cer -inform DER -outform PEM -out cert.pem
    // $ openssl pkcs12 -in key.p12 -out key.pem -nodes



    // This function is called when a token is replaced on a device - normally
    // this should not happen, but if it does we should take action on it
    _replaceToken = function(currentToken, newToken) {
        // console.log('Replace token: ' + currentToken + ' -- ' + newToken);
        // If the server gets a token event its passing in the current token and
        // the new value - if new value is undefined this empty the token
        self.emit('token', currentToken, newToken);
    };

    // Rig the removeToken callback
    _removeToken = function(token) {
        // console.log('Remove token: ' + token);
        // Invalidate the token
        self.emit('token', token, null);
    };


    if (options.apn) {
        if (Push.debug) console.log('Push: APN configured');

        // Allow production to be a general option for push notifications
        if (options.production)
          options.apn.production = true;

        // We check the apn gateway i the options, we could risk shipping
        // server into production while using the production configuration.
        // On the other hand we could be in development but using the production
        // configuration. And finally we could have configured an unknown apn
        // gateway (this could change in the future - but a warning about typos
        // can save hours of debugging)
        //
        // Warn about gateway configurations - it's more a guide
        if (options.apn.gateway) {

            if (options.apn.gateway == 'gateway.sandbox.push.apple.com') {
                // Using the development sandbox
                console.warn('WARNING: Push APN is in development mode');
            } else if (options.apn.gateway == 'gateway.push.apple.com') {
                // In production - but warn if we are running on localhost
                if (/http:\/\/localhost/.test(Meteor.absoluteUrl())) {
                    console.warn('WARNING: Push APN is configured to production mode - but server is running from localhost');
                }
            } else {
                // Warn about gateways we dont know about
                console.warn('WARNING: Push APN unkown gateway "' + options.apn.gateway + '"');
            }

        } else {
            if (options.apn.production) {
                if (/http:\/\/localhost/.test(Meteor.absoluteUrl())) {
                    console.warn('WARNING: Push APN is configured to production mode - but server is running from localhost');
                }
            } else {
                console.warn('WARNING: Push APN is in development mode');
            }
        }

        // Check certificate data
        if (!options.apn['certData'] || !options.apn['certData'].length)
            console.error('ERROR: Push server could not find certData');

        // Check key data
        if (!options.apn['keyData'] || !options.apn['keyData'].length)
            console.error('ERROR: Push server could not find keyData');

        // Rig apn connection
        var apn = Npm.require('apn');
        var apnConnection = new apn.Connection( options.apn );


        // XXX: should we do a test of the connection? It would be nice to know
        // That the server/certificates/network are correct configured

        // apnConnection.connect().then(function() {
        //     console.info('CHECK: Push APN connection OK');
        // }, function(err) {
        //     console.warn('CHECK: Push APN connection FAILURE');
        // });
        // Note: the above code spoils the connection - investigate how to
        // shutdown/close it.

        // (cert.pem and key.pem)
        self.sendAPN = function(from, userToken, title, text, count, priority) {
            // console.log('sendAPN', from, userToken, title, text, count, priority);
            priority = (priority || priority === 0)? priority : 10;

            var myDevice = new apn.Device(userToken);

            var note = new apn.Notification();

            note.expiry = Math.floor(Date.now() / 1000) + 3600; // Expires 1 hour from now.
            note.badge = count;
            note.sound = ""; // XXX: Does this work?
            note.alert = text;
            note.payload = {'messageFrom': from };
            note.priority = priority;

            // console.log('I:Send message to: ' + userToken + ' count=' + count);

            apnConnection.pushNotification(note, myDevice);

        };


        var initFeedback = function() {
            var apn = Npm.require('apn');
            // console.log('Init feedback');
            var feedbackOptions = {
                "batchFeedback": true,
                "interval": 1000,
                'address': 'feedback.push.apple.com'
            };

            var feedback = new apn.Feedback(feedbackOptions);
            feedback.on("feedback", function(devices) {
                devices.forEach(function(item) {
                    // Do something with item.device and item.time;
                    // console.log('A:PUSH FEEDBACK ' + item.device + ' - ' + item.time);
                    // The app is most likely removed from the device, we should
                    // remove the token
                    _removeToken({ apn: item.device});
                });
            });
        };

        // Init feedback from apn server
        // This will help keep the appCollection up-to-date, it will help update
        // and remove token from appCollection.
        initFeedback();

    } // EO ios options

    if (options.gcm && options.gcm.apiKey) {
        if (Push.debug) console.log('GCM configured');
        self.sendGCM = function(from, userTokens, title, text, count, priority) {
            // Make sure userTokens are an array of strings
            if (userTokens === ''+userTokens) userTokens = [userTokens];

            // Check if any tokens in there to send
            if (!userTokens.length) {
                if (Push.debug) console.log('sendGCM no push tokens found');
                return;
            }

            if (Push.debug)
                console.log('sendGCM', from, userTokens, title, text, count, priority);

            var gcm = Npm.require('node-gcm');
            var Fiber = Npm.require('fibers');

            //var message = new gcm.Message();
            var message = new gcm.Message({
                collapseKey: from,
            //    delayWhileIdle: true,
            //    timeToLive: 4,
            //    restricted_package_name: 'dk.gi2.app'
                data: {
                    title: title,
                    message: text,
                    msgcnt: count
                }
            });

            if (Push.debug) console.log('Create GCM Sender using "' + options.gcm.apiKey + '"');
            var sender = new gcm.Sender(options.gcm.apiKey);

            _.each(userTokens, function(value, key) {
                if (Push.debug) console.log('A:Send message to: ' + value + ' count=' + count);
            });

            /*message.addData('title', title);
            message.addData('message', text);
            message.addData('msgcnt', '1');
            message.collapseKey = 'sitDrift';
            message.delayWhileIdle = true;
            message.timeToLive = 3;*/

            // /**
            //  * Parameters: message-literal, userTokens-array, No. of retries, callback-function
            //  */

            var userToken = (userTokens.length === 1)?userTokens[0]:null;

            sender.send(message, userTokens, 5, function (err, result) {
                if (err) {
                    if (Push.debug) console.log('ANDROID ERROR: result of sender: ' + result);
                } else {
                    if (Push.debug) console.log('ANDROID: Result of sender: ' + JSON.stringify(result));
                    if (result.canonical_ids === 1 && userToken) {

                        // This is an old device, token is replaced
                        Fiber(function(self) {
                            // Run in fiber
                            try {
                                self.callback(self.oldToken, self.newToken);
                            } catch(err) {

                            }

                        }).run({
                            oldToken: { gcm: userToken },
                            newToken: { gcm: result.results[0].registration_id },
                            callback: _replaceToken
                        });
                        //_replaceToken({ gcm: userToken }, { gcm: result.results[0].registration_id });

                    }
                    // We cant send to that token - might not be registred
                    // ask the user to remove the token from the list
                    if (result.failure !== 0 && userToken) {

                        // This is an old device, token is replaced
                        Fiber(function(self) {
                            // Run in fiber
                            try {
                                self.callback(self.token);
                            } catch(err) {

                            }

                        }).run({
                            token: { gcm: userToken },
                            callback: _removeToken
                        });
                        //_replaceToken({ gcm: userToken }, { gcm: result.results[0].registration_id });

                    }

                }
            });
            // /** Use the following line if you want to send the message without retries
            // sender.sendNoRetry(message, userTokens, function (result) {
            //     console.log('ANDROID: ' + JSON.stringify(result));
            // });
            // **/
        }; // EO sendAndroid

    } // EO Android

    // Universal send function
    var _querySend = function(from, query, title, text, count, priority) {

      var countApn = [];
      var countGcm = [];

        Push.appCollection.find(query).forEach(function(app) {

          if (Push.debug) console.log('send to token', app.token);

            if (app.token.apn) {
              countApn.push(app._id);
                // Send to APN
                if (self.sendAPN) self.sendAPN(from, app.token.apn, title, text, count, priority);

            } else if (app.token.gcm) {
              countGcm.push(app._id);

                // Send to GCM
                // We do support multiple here - so we should construct an array
                // and send it bulk - Investigate limit count of id's
                if (self.sendGCM) self.sendGCM(from, app.token.gcm, title, text, count, priority);

            } else {
                throw new Error('Push.send got a faulty query');
            }

        });

        if (Push.debug) console.log('Push: Sent message "' + title + '" to ' + countApn.length + ' ios apps ' + countGcm.length + ' android apps');

        return {
          apn: countApn,
          gcm: countGcm
        };
    };

    self.serverSend = function(options) {
      options = options || { count: 0 };
      var query;

      // Check basic options
      if (options.from !== ''+options.from)
        throw new Error('Push.send: option "from" not a string');

      if (options.title !== ''+options.title)
        throw new Error('Push.send: option "title" not a string');

      if (options.text !== ''+options.text)
        throw new Error('Push.send: option "text" not a string');

      if (options.token || options.tokens) {

        // The user set one token or array of tokens
        var tokenList = (options.token)? [options.token] : options.tokens;

        if (Push.debug) console.log('Push: Send message "' + options.title + '" via token(s)', tokenList);

        query = {
          $or: [
              // XXX: Test this query: can we hand in a list of push tokens?
              { token: { $in: tokenList } },
              // XXX: Test this query: does this work on app id?
              { $and: [
                  { _in: { $in: tokenList } }, // one of the app ids
                  { $or: [
                      { 'token.apn': { $exists: true }  }, // got apn token
                      { 'token.gcm': { $exists: true }  }  // got gcm token
                  ]}
              ]}
          ]
        };

      } else if (options.query) {

        if (Push.debug) console.log('Push: Send message "' + options.title + '" via query', options.query);

        query = {
          $and: [
              options.query, // query object
              { $or: [
                  { 'token.apn': { $exists: true }  }, // got apn token
                  { 'token.gcm': { $exists: true }  }  // got gcm token
              ]}
          ]
        };
      }


      if (query) {

        // Convert to querySend and return status
        return _querySend(options.from, query, options.title, options.text, options.count, options.priority)

      } else {
        throw new Error('Push.send: please set option "token"/"tokens" or "query"');
      }

    };

    var isSendingNotification = false;

    Meteor.setInterval(function() {

        if (!isSendingNotification) {
            // Set send fence
            isSendingNotification = true;

            // Find one notification
            var notification = Push.notifications.findOne({ sent : { $ne: true } }, { sort: { createdAt: 1 } });

            // Check if we got any notifications to send
            if (notification) {

                // Send the notification
                var result = Push.serverSend(notification);

                if (!options.keepNotifications) {
                    // Pr. Default we will remove notifications
                    Push.notifications.remove({ _id: notification._id });
                } else {

                    // Update the notification
                    Push.notifications.update({ _id: notification._id }, {
                        $set: {
                            sent: true,
                            sentAt: new Date(),
                            count: result
                        }
                    });

                }

                // Emit the send
                self.emit('send', { notification: notification._id, result: result });
            }

            // Remove the send fence
            isSendingNotification = false;
        }
    }, options.sendInterval || 15000); // Default every 15'the sec

};
